### PR TITLE
CW L0 attack

### PR DESCRIPTION
**Pull request by carlini**
This branch implements the CW L0 attack. It is implemented on top of the L2 attack, so some generalizations are required there.

### CODE REVIEW

**Comment by carlini** [comment]
mnist_tutorial_tf is failing erroneously, it seems (I didn't touch that file). I'm not sure why, but the python 3.5 build is timing out.

**Comment by npapernot** [comment]
I restarted it but the tests are still failing (it appears one test is not found)

File: cleverhans/attacks_tf.py [hunk 1]
```diff
@@ -519,6 +523,7 @@ def __init__(self, sess, model, batch_size, confidence,
         self.clip_min = clip_min
         self.clip_max = clip_max
         self.model = model
+        self.extension = extension
```
Comment (jmgilmer) [inline cleverhans/attacks_tf.py]: Add what extension does to the doc string.

File: cleverhans/attacks_tf.py [hunk 2]
```diff
@@ -606,7 +631,7 @@ def attack(self, imgs, targets):
                                        targets[i:i + self.batch_size]))
         return np.array(r)
 
-    def attack_batch(self, imgs, labs):
+    def attack_batch(self, imgs, labs, mask=None):
```
Comment (jmgilmer) [inline cleverhans/attacks_tf.py]: What will mask be used for? Also I don't see a reference to a mask described in the paper, I assume it's used in the case the attacker only wants to perturb a subset of the image.

Comment (jmgilmer) [reply cleverhans/attacks_tf.py]: How come attack_batch uses the mask argument but attack doesn't?

Comment (carlini) [reply cleverhans/attacks_tf.py]: It's only needed for the L0 attack algorithm, so I only put it there for calling into it directly. I don't want to directly expose it to the outside world.

File: cleverhans/attacks_tf.py [hunk 3]
```diff
@@ -725,9 +757,162 @@ def compare(x, y):
...
+            self.l2_attack.source_image = np.copy(prev)
+            while const < self.largest_const:
+                # try solving for each value of the constant
+                self.l2_attack.initial_const = const
+                res = self.l2_attack.attack_batch(np.copy([instance]),
+                                                  np.array([target]),
+                                                  mask=np.array([valid]))
+                if res is not None:
+                    break
+                const *= self.const_factor
```
Comment (jmgilmer) [inline cleverhans/attacks_tf.py]: Just double checking that we dont' want to initialize const to self.initial_const before entering this inner loop.

File: cleverhans/attacks_tf.py [hunk 4]
```diff
@@ -725,9 +757,162 @@ def compare(x, y):
...
+            _logger.debug("Next iteration; {} fixed values."
+                          .format(equal_count))
+
+            orig_shape = valid.shape
+            valid = valid.flatten()
+            totalchange = abs(nimg[0]-instance)*np.abs(gradientnorm[0])
+            totalchange = totalchange.flatten()
+
+            # set some of the pixels to 0 depending on their total change
+            did = 0
```
Comment (jmgilmer) [inline cleverhans/attacks_tf.py]: Is there a more descriptive name for this variable other than did? I'm not sure what this is tracking by just looking at the code.

File: cleverhans/attacks_tf.py [hunk 5]
```diff
@@ -725,9 +757,162 @@ def compare(x, y):
...
+            for e in np.argsort(totalchange):
+                if np.all(valid[e]):
+                    did += 1
+                    valid[e] = 0
+
+                    if totalchange[e] > .01:
```
Comment (jmgilmer) [inline cleverhans/attacks_tf.py]: Maybe .01 should be a variable call MAX_PIXEL_CHANGE?

File: cleverhans/attacks_tf.py [hunk 6]
```diff
@@ -725,9 +757,162 @@ def compare(x, y):
...
+                        # if this pixel changed a lot, skip
+                        break
+                    if did >= .3*equal_count**.5:
```
Comment (jmgilmer) [inline cleverhans/attacks_tf.py]: Similar to the .01, what is special about .3 here?

File: cleverhans_tutorials/mnist_tutorial_cw.py [hunk 7]
```diff
@@ -105,100 +106,88 @@ def mnist_tutorial_cw(train_start=0, train_end=60000, test_start=0,
...
-        adv_inputs = adv_inputs.reshape(
-            (source_samples * nb_classes, img_rows, img_cols, 1))
-        adv_ys = np.array([one_hot] * source_samples,
-                          dtype=np.float32).reshape((source_samples *
-                                                     nb_classes, nb_classes))
+        # Initialize our array for grid visualization
+        grid_shape = (nb_classes, nb_classes, img_rows, img_cols, channels)
+        grid_viz_data = np.zeros(grid_shape, dtype='f')
+
+        one_hot = np.zeros((10, 10))
```
Comment (jmgilmer) [inline cleverhans_tutorials/mnist_tutorial_cw.py]: Consider using np.eye here.

File: cleverhans/attacks_tf.py (hunk 1 above)
Comment (carlini) [reply cleverhans/attacks_tf.py]: Done

File: cleverhans/attacks_tf.py (hunk 3 above)
Comment (carlini) [reply cleverhans/attacks_tf.py]: Nope, want it to stay at the old value.

File: cleverhans/attacks_tf.py (hunk 5 above)
Comment (carlini) [reply cleverhans/attacks_tf.py]: Yeah. I will make this change.

File: cleverhans/attacks_tf.py (hunk 6 above)
Comment (carlini) [reply cleverhans/attacks_tf.py]: Also will add this as a named constant.

**Comment by catherio** [comment]
@jmgilmer @carlini - Looks like this has been sitting for a month or so, any updates?

**Comment by carlini** [comment]
I will address these comments next week or so -- have a few things I need finish over the next couple of days.

**Comment by catherio** [comment]
@carlini Any updates? No worries if not, I'm just doing a routine pass through all our pull requests to check their status.

**Comment by carlini** [comment]
Huh, I thought I pushed a few days ago but I guess it didn't go. Let me see.

**Comment by carlini** [comment]
It looks like the build is broken for different reasons than this PR.

**Comment by catherio** [comment]
I'm starting to take a look now: #339 and #340

**Comment by catherio** [comment]
@carlini build is fixed on master now as of PR #343 , you should be able to rebase

**Comment by catherio** [comment]
@carlini - wanted to make sure you saw my earlier message. the build is passing now, so if you want to rebase and resolve the conflict, this should be ready to merge.

**Comment by carlini** [comment]
I'm going to close this until I have time to fix everything. I'll submit a new PR once it's ready.

**Pull request closed without merging**